hbase-server/src/main/java/org/apache/hadoop/hbase/master/MasterStateStore.java

```diff
@@ -0,0 +1,103 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.master;
+
+import static org.apache.hadoop.hbase.master.region.MasterRegionFactory.STATE_FAMILY;
+
+import java.io.IOException;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.master.region.MasterRegion;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hadoop.hbase.zookeeper.ZKUtil;
+import org.apache.hadoop.hbase.zookeeper.ZKWatcher;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A state storage which stores the state in master local region.
+ * <p/>
+ * We used to store some persistent state on zookeeper, so here we provide the ability to migrate
+ * the state from zookeeper.
+ * <p/>
+ * Since parsing the state may introduce some extra overhead, we make this class abstract and the
+ * get and set state methods protected. Sub classes should store their state in decoded format to
+ * save the extra parsing overhead.
+ */
+@InterfaceAudience.Private
+public abstract class MasterStateStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MasterStateStore.class);
+
+  private static final byte[] QUALIFIER = Bytes.toBytes("d");
+
+  private final MasterRegion masterRegion;
+
+  private final byte[] stateName;
+
+  private Result get() throws IOException {
+    return masterRegion.get(new Get(stateName).addColumn(STATE_FAMILY, QUALIFIER));
+  }
+
+  private void update(byte[] s) throws IOException {
+    masterRegion.update(r -> r.put(new Put(stateName).addColumn(STATE_FAMILY, QUALIFIER, s)));
+  }
+
+  private byte[] migrate(ZKWatcher watcher, String zkPath) throws KeeperException, IOException {
+    byte[] zkData = ZKUtil.getDataNoWatch(watcher, zkPath, null);
+    if (zkData == null || zkData.length == 0) {
+      return null;
+    }
+    update(zkData);
+    return zkData;
+  }
+
+  private void tryMigrate(ZKWatcher watcher, String zkPath) throws IOException, KeeperException {
+    Result result = masterRegion.get(new Get(stateName).addColumn(STATE_FAMILY, QUALIFIER));
+    if (result.isEmpty()) {
+      // migrate
+      migrate(watcher, zkPath);
+    }
+    // we may fail in the middle so even if the value is available in master local region, we may
+    // still leave a znode on zookeeper, so always try to delete the znode here since it is not very
+    // expensive
+    try {
+      ZKUtil.deleteNodeFailSilent(watcher, zkPath);
+    } catch (Exception e) {
+      LOG.warn("failed to delete migrated zk state node {}, ignore and continue", zkPath);
+    }
+  }
```

Comment:
can you move these private methods below the protected methods so it reads a little more naturally?